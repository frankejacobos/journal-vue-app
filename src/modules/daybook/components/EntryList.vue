<template>
  <div class="entry-list-container">
    <div class="px-2 pt-2">
      <input
        type="text"
        class="form-control"
        placeholder="Buscar algo"
        v-model="search"
      />
    </div>
    <div class="mt-2 d-flex flex-column">
      <button
        class="btn btn-primary m-3"
        @click="$router.push({ name: 'entry', params: { id: 'new' } })"
      >
        <i class="fa fa-plus-circle"></i>
        Nueva entrada
      </button>
    </div>
    <div class="entry-scrollarea">
      <entry-item
        v-for="entry in entriesByTerm"
        :key="entry.id"
        :entry="entry"
      ></entry-item>
    </div>
  </div>
</template>

<script>
import { defineAsyncComponent } from "vue";
import { mapGetters } from "vuex";

export default {
  components: {
    EntryItem: defineAsyncComponent(() => import("./EntryItem.vue")),
  },
  computed: {
    ...mapGetters("journal", ["getEntriesByTerm"]),
    entriesByTerm() {
      return this.getEntriesByTerm(this.search);
    },
  },
  data() {
    return {
      search: "",
    };
  },
};
</script>

<style scoped>
.entry-list-container {
  height: calc(100vh - 66px);
}

.entry-scrollarea {
  height: calc(100vh - 64px - 48px);
  overflow: scroll;
}
</style>
