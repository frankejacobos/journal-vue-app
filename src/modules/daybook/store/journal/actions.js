import journalApi from "@/api/journalApi";

export const createEntry = async ({ commit }, entry) => {
  const { date, text, picture } = entry;
  const dataToSave = { date, text, picture };
  const { data } = await journalApi.post("/entries.json", dataToSave);
  commit("addEntry", { ...dataToSave, id: data.name });
  return data.name;
};

export const readEntries = async ({ commit }) => {
  const { data } = await journalApi.get("/entries.json");
  let entries = [];
  if (data) {
    for (let id of Object.keys(data)) {
      entries.push({
        id,
        ...data[id],
      });
    }
  }
  commit("setEntries", entries);
};

export const updateEntry = async ({ commit }, entry) => {
  const { date, text, picture } = entry;
  const dataToSave = { date, text, picture };
  await journalApi.put(`/entries/${entry.id}.json`, dataToSave);
  commit("updateEntry", { ...entry });
};

export const deleteEntry = async ({ commit }, id) => {
  await journalApi.delete(`/entries/${id}.json`);
  commit("deleteEntry", id);
};

export const clearEntries = ({ commit }) => {
  commit("clearEntries");
};
