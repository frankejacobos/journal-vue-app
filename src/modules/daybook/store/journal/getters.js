// export const getter = (state) => {
//   return state.property;
// };

export const getEntriesByTerm =
  (state) =>
  (term = "") => {
    return term.length > 0
      ? state.entries.filter((entry) =>
          entry.text.toLowerCase().includes(term.toLowerCase())
        )
      : state.entries;
  };

export const getEntryById =
  (state) =>
  (id = "") => {
    const entry = state.entries.find((entry) => entry.id === id);
    return entry ? { ...entry } : undefined;
  };
