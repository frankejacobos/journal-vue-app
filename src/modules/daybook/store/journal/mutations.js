export const setEntries = (state, entries) => {
  state.entries = [...state.entries, ...entries];
  state.isLoading = false;
};

export const addEntry = (state, entry) => {
  state.entries = [entry, ...state.entries];
};

export const updateEntry = (state, entry) => {
  const index = state.entries.findIndex((e) => e.id === entry.id);
  state.entries[index] = entry;
};

export const deleteEntry = (state, entryId) => {
  state.entries = state.entries.filter((e) => e.id !== entryId);
};
