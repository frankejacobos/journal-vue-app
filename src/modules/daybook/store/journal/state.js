const date1 = new Date();
date1.setDate(date1.getDate() - 15);
const date2 = new Date();
date2.setDate(date2.getDate() - 5);
const date3 = new Date();
date3.setDate(date3.getDate() - 1);

const dates = [date1, date2, date3];

export default () => ({
  isLoading: true,
  entries: [
    {
      id: "1",
      date: dates[0].toDateString(),
      picture: null,
      text: "Officia irure id excepteur labore. Reprehenderit esse aliqua nulla amet reprehenderit culpa aute dolore consequat aute consequat nostrud. Culpa id tempor sint voluptate anim reprehenderit fugiat.",
    },
    {
      id: "2",
      date: dates[1].toDateString(),
      picture: null,
      text: "Ut laborum aute proident laboris occaecat do quis nulla. Ut nostrud enim do ipsum ut aliqua reprehenderit ipsum labore adipisicing do mollit esse culpa. Minim occaecat non nostrud dolor incididunt enim. Sit quis elit sit esse est ipsum commodo eu sint do ipsum mollit laborum. Nulla ullamco est voluptate et nulla sunt.",
    },
    {
      id: "3",
      date: dates[2].toDateString(),
      picture: null,
      text: "Eiusmod commodo do nisi est pariatur. Id dolor fugiat ut nulla. Tempor veniam commodo cillum proident officia do pariatur proident eu culpa nostrud ex amet. Enim Lorem cupidatat laboris in irure Lorem id ex esse aliquip. Ea ullamco ut adipisicing dolore sunt sit aliquip cillum pariatur et adipisicing cupidatat cillum. Do eiusmod minim dolore esse.",
    },
  ],
});
