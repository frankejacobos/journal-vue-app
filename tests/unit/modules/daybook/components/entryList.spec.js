import { createStore } from "vuex";
import { shallowMount } from "@vue/test-utils";
import { journalState } from "@/../tests/unit/mocks/test-journal-state";
import EntryList from "@/modules/daybook/components/EntryList.vue";
import journal from "@/modules/daybook/store/journal";

const createVuexStore = (state) =>
  createStore({
    modules: {
      journal: {
        ...journal,
        state: { ...state },
      },
    },
  });
describe("Pruebas en el componente EntryList", () => {
  const store = createVuexStore(journalState);
  const mockRouter = { push: jest.fn() };
  let wrapper;
  beforeEach(() => {
    jest.clearAllMocks();
    wrapper = shallowMount(EntryList, {
      global: {
        mocks: { $router: mockRouter },
        plugins: [store],
      },
    });
  });
  test("debe de llamar getEntriesByTerm sin termino y retornar 2 entries", () => {
    expect(wrapper.findAll("entry-item-stub").length).toBe(2);
    expect(wrapper.html()).toMatchSnapshot();
  });
  test("debe de llamar getEntriesByTerm con termino y retornar 1 entry", async () => {
    await wrapper.find("input").setValue("Hola");
    expect(wrapper.findAll("entry-item-stub").length).toBe(1);
    expect(wrapper.html()).toMatchSnapshot();
  });
  test("debe de llamar getEntriesByTerm con termino y retornar 0 entry", async () => {
    await wrapper.find("input").setValue("Hola mndo");
    expect(wrapper.findAll("entry-item-stub").length).toBe(0);
    expect(wrapper.html()).toMatchSnapshot();
  });
  test('debe redireccionar a /new cuando se haga click en el boton "new"', () => {
    wrapper.find("button").trigger("click");
    expect(mockRouter.push).toHaveBeenCalledWith({
      name: "entry",
      params: { id: "new" },
    });
  });
});
