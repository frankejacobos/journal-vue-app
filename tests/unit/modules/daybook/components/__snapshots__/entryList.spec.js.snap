// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente EntryList debe de llamar getEntriesByTerm con termino y retornar 0 entry 1`] = `
<div class="entry-list-container">
  <div class="px-2 pt-2"><input type="text" class="form-control" placeholder="Buscar algo"></div>
  <div class="mt-2 d-flex flex-column"><button class="btn btn-primary m-3"><i class="fa fa-plus-circle"></i> Nueva entrada </button></div>
  <div class="entry-scrollarea"></div>
</div>
`;

exports[`Pruebas en el componente EntryList debe de llamar getEntriesByTerm con termino y retornar 1 entry 1`] = `
<div class="entry-list-container">
  <div class="px-2 pt-2"><input type="text" class="form-control" placeholder="Buscar algo"></div>
  <div class="mt-2 d-flex flex-column"><button class="btn btn-primary m-3"><i class="fa fa-plus-circle"></i> Nueva entrada </button></div>
  <div class="entry-scrollarea">
    <entry-item-stub entry="[object Object]"></entry-item-stub>
  </div>
</div>
`;

exports[`Pruebas en el componente EntryList debe de llamar getEntriesByTerm sin termino y retornar 2 entries 1`] = `
<div class="entry-list-container">
  <div class="px-2 pt-2"><input type="text" class="form-control" placeholder="Buscar algo"></div>
  <div class="mt-2 d-flex flex-column"><button class="btn btn-primary m-3"><i class="fa fa-plus-circle"></i> Nueva entrada </button></div>
  <div class="entry-scrollarea">
    <entry-item-stub entry="[object Object]"></entry-item-stub>
    <entry-item-stub entry="[object Object]"></entry-item-stub>
  </div>
</div>
`;
